topgun/k8s/https_web_tls_termination_test.go

```diff
@@ -0,0 +1,149 @@
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gbytes"
+	"github.com/onsi/gomega/gexec"
+	"github.com/square/certstrap/pkix"
+	"io/ioutil"
+	"net"
+	"os"
+	"time"
+)
+
+var _ = Describe("Web HTTP or TLS termination at web node", func() {
+	var (
+		proxySession  *gexec.Session
+		atcEndpoint   string
+		chartConfig   []string
+		proxyPort     string
+		proxyProtocol string
+	)
+
+	JustBeforeEach(func() {
+		setReleaseNameAndNamespace("wtt")
+		deployConcourseChart(releaseName,
+			chartConfig...)
+
+		waitAllPodsInNamespaceToBeReady(namespace)
+
+		By("Creating the web proxy")
+		proxySession, atcEndpoint = startPortForwardingWithProtocol(namespace, "service/"+releaseName+"-web", proxyPort, proxyProtocol)
+
+	})
+	AfterEach(func() {
+		helmDestroy(releaseName)
+		Wait(Start(nil, "kubectl", "delete", "namespace", namespace, "--wait=false"))
+		Wait(proxySession.Interrupt())
+	})
+
+	Context("configure helm chart for tls termination at web", func() {
+		var caCertFile *os.File
+		BeforeEach(func() {
+			CACert, serverKey, serverCert := generateKeyPairWithCA()
+			CACertBytes, err := CACert.Export()
+			Expect(err).NotTo(HaveOccurred())
+			caCertFile, err = ioutil.TempFile("", "ca")
+			caCertFile.Write(CACertBytes)
+			caCertFile.Close()
+
+			serverKeyBytes, err := serverKey.ExportPrivate()
+			Expect(err).NotTo(HaveOccurred())
+
+			serverCertBytes, err := serverCert.Export()
+			Expect(err).NotTo(HaveOccurred())
+			//Expect(err).ToNot(HaveOccurred())
+
+			chartConfig = generateChartConfig(
+				"--set=concourse.web.externalUrl=https://test.com",
+				"--set=concourse.web.tls.enabled=true",
+				"--set=secrets.webTlsCert="+string(serverCertBytes),
+				"--set=secrets.webTlsKey="+string(serverKeyBytes),
+			)
+			proxyPort = "443"
+			proxyProtocol = "https"
+		})
+		AfterEach(func() {
+			os.Remove(caCertFile.Name())
+		})
+		It("fly login succeeds when using the correct CA and host", func() {
+			By("Logging in")
+			sess := fly.Start("login", "-u", "test", "-p", "test", "--ca-cert", caCertFile.Name(), "-c", atcEndpoint)
+			<-sess.Exited
+			Expect(sess.ExitCode()).To(Equal(0))
+		})
+		It("fly login fails when NOT using the correct CA", func() {
+			By("Logging in")
+			sess := fly.Start("login", "-u", "test", "-p", "test", "--ca-cert", "certs/wrong-ca.crt", "-c", atcEndpoint)
+			<-sess.Exited
+			Expect(sess.ExitCode()).ToNot(Equal(0))
+			Expect(sess.Err).To(gbytes.Say(`x509: certificate signed by unknown authority`))
+		})
+	})
+	Context("DON'T configure tls termination at web in helm chart", func() {
+		BeforeEach(func() {
+			chartConfig = generateChartConfig("--set=concourse.web.externalUrl=http://test.com",
+				"--set=concourse.web.tls.enabled=false")
+			proxyPort = "8080"
+			proxyProtocol = "http"
+		})
+		It("fly login succeeds when connecting to web over http", func() {
+			By("Logging in")
+			fly.Login("test", "test", atcEndpoint)
+		})
+	})
+})
+
+var _ = Describe("When TLS is not configured correctly", func() {
```

Comment:
Hey,

@taylorsilva and I think that this should instead of being a sibling, be a child of the previous `Describe` as this is within the same thing being tested (`atc` terminating TLS connections). Perhaps we could even rename this to a `Context`. Wdyt?

Thanks!